File: frontend/src/types/reports.ts
export type LowStockItem = {
  id: number;
  name: string;
  barcode: string;
  sku: string;
  quantity: number;
  min_quantity: number;
  shortfall: number;
  image_url: string | null;
};

export type LowStockResponse = {
  count: number;
  results: LowStockItem[];
};

File: backend/apps/products/reports.py
from django.db.models import F, IntegerField, QuerySet, Value
from django.db.models.functions import Coalesce

from apps.products.models import Product


def list_low_stock_products() -> QuerySet[Product]:
    return (
        Product.objects.only(
            "id",
            "name",
            "barcode",
            "sku",
            "quantity",
            "min_quantity",
            "image",
            "created_at",
            "updated_at",
        )
        .filter(quantity__lte=F("min_quantity"))
        .annotate(shortfall=Coalesce(F("min_quantity") - F("quantity"), Value(0), output_field=IntegerField()))
        .order_by("quantity", "name", "barcode")
    )
